File: college/algorithms/4 cryptology/1_rsa/rsa.py
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.primitives import serialization

# Generate RSA key pair
def generate_key_pair():
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=2048,
        backend=default_backend()
    )
    public_key = private_key.public_key()
    return private_key, public_key

# Encrypt message using RSA public key
def encrypt_message(message, public_key):
    encrypted = public_key.encrypt(
        message.encode(),
        padding.PKCS1v15()
    )
    return encrypted

# Decrypt message using RSA private key
def decrypt_message(encrypted_message, private_key):
    decrypted = private_key.decrypt(
        encrypted_message,
        padding.PKCS1v15()
    )
    return decrypted.decode()

# Example usage
private_key, public_key = generate_key_pair()

message = "Hello, this is a secret message!"
print("Original Message:", message)

# Encrypt the message with the public key
encrypted_message = encrypt_message(message, public_key)
print("Encrypted Message:", encrypted_message)

# Decrypt the message with the private key
decrypted_message = decrypt_message(encrypted_message, private_key)
print("Decrypted Message:", decrypted_message)

File: college/day-department/algorithms/4 cryptology/1_rsa/rsa.cs
using System;
using System.Security.Cryptography;
using System.Text;

class RSAExample
{
    static void Main(string[] args)
    {
        // Generate RSA key pair
        using (RSACryptoServiceProvider rsa = new RSACryptoServiceProvider(2048))
        {
            try
            {
                // Export public key
                string publicKey = rsa.ToXmlString(false);
                // Export private key
                string privateKey = rsa.ToXmlString(true);

                string message = "Hello, this is a secret message!";
                Console.WriteLine("Original Message: " + message);

                // Encrypt message using public key
                byte[] encryptedBytes = Encrypt(Encoding.UTF8.GetBytes(message), publicKey);
                string encryptedMessage = Convert.ToBase64String(encryptedBytes);
                Console.WriteLine("Encrypted Message: " + encryptedMessage);

                // Decrypt message using private key
                byte[] decryptedBytes = Decrypt(Convert.FromBase64String(encryptedMessage), privateKey);
                string decryptedMessage = Encoding.UTF8.GetString(decryptedBytes);
                Console.WriteLine("Decrypted Message: " + decryptedMessage);
            }
            catch (CryptographicException e)
            {
                Console.WriteLine(e.Message);
            }
        }
    }

    // Encrypt message using RSA public key
    static byte[] Encrypt(byte[] data, string publicKey)
    {
        using (RSACryptoServiceProvider rsa = new RSACryptoServiceProvider())
        {
            rsa.FromXmlString(publicKey);
            return rsa.Encrypt(data, false);
        }
    }

    // Decrypt message using RSA private key
    static byte[] Decrypt(byte[] data, string privateKey)
    {
        using (RSACryptoServiceProvider rsa = new RSACryptoServiceProvider())
        {
            rsa.FromXmlString(privateKey);
            return rsa.Decrypt(data, false);
        }
    }
}
